flips/20211007-transaction-fees.md

```diff
@@ -0,0 +1,264 @@
+
+As discussed below, in scenario 1. the transaction shouldn't have passed the access node in the first place. However since it has and there is no way to charge the payer, the access node account is charged with the transaction fees. The transaction fees are computed with the execution effort set to 0.
+
+In the second scenario the payer is charged for the transaction fees with the execution effort set to 0.
+
+In the third case, the fees are charged normally with the execution effort set to the actual usage before the error was produced.
+
+Finally in the last case the payer is charged with the transaction fees where the execution effort is set to the execution effort limit.
+
+### Preventing malicious actors
+
+The purpose of fees is to prevent malicious actors sending transactions with the purpose of overloading the network. To achieve that, we have to check that the payer of a transaction has the funds to pay for that transaction as soon as possible.
+
+Tho achieve this the following checks should happen on the access node, before the transaction is even included in a collection: 
+- Is the payers signature valid.
+- Can the payer pay for the transaction. The payers default FLOW vault should have at least <img src="https://render.githubusercontent.com/render/math?math=F%5E%5Ctext%7Bmax%7D"> FLOW. <img src="https://render.githubusercontent.com/render/math?math=F%5E%7Bmax%7D%20%3D%20f%20(%20E_i%20%2B%20E_e%5E%7Bmax%7D%20)">
+
+It might still happen that this transaction will get included in a collection, either due to a bug or malicious behaviour from the access node, or due to the access node not having up to date information of the balance of the payer. In case does still get included in a collection, even though it does not meet the condition, this check should be repeated in the execution node, and in case the check fails on the execution node, the transaction fees should be deducted from the access nodes account instead of the payers account.
+
+### Calibrating the parameters
+
+When turning on variable transaction fees the goal is to have a smooth transition to the new system. To help facilitate that the cost of transaction fees should be set so that ~95% of transactions should actually pay less transaction fees with variable transaction fees turned on.
+
+To achieve this the 6 different parts of variable transaction fees need to be calibrated accordingly.
+
+- <img src="https://render.githubusercontent.com/render/math?math=E_i"> Inclusion effort would be the byte size of a transaction plus a constant inclusion effort that would be picked by using linear regression on the impact of different size transactions (with the same execution effort) on the network, and extrapolating for a 0 size transaction.
+- <img src="https://render.githubusercontent.com/render/math?math=f_i">: Effort cost factor would be picked so that for 95% of transactions that have been seen so far the inclusion fees would cost less then `TC/2`.
```

Comment:
regarding parameter calibration: 
* so we have effort, e.g. we could define that every 1kb transaction size equals 1 unit of inclusion effort
* and then we have a effort cost, i.e. how much does 1kb transaction size costs (?)

In economy, you often draw a price vs quantity diagram: 
<img width="1239" alt="Screen Shot 2021-11-25 at 8 59 49 PM" src="https://user-images.githubusercontent.com/26322303/143530448-c7a622a4-e1c4-437c-a08d-6429bf357441.png">

I have introduced several quantities here:
* `E_95` the 95% effort percentile of current transactions (i.e. their size in kb)
* `F_c` the current (fixed) transaction fee
* Furthermore, lets consider another transaction A with different effort (This could be a very small transaction with close to minimal effort `E_min`, as shown above) or a larger transactions (illustrated below).
* Assuming you desire a linear cost-effort relationship:
   * this is a linear function which is uniquely defined by a single point `X` and a slope `α`
   * we already have one point fixed: `X_95 = (E_95, F_c)`
   * for another transaction with effort `E_A` we still lack a meaningful price value `F_A`
* So how do we determine `F_A`? 
   * This is where regression comes into play, but I don't think the ordinate intersection point is particularly helpful here (you call this the "0-size transaction"). As an illustration, consider the two different price curves `p^(1)` vs `p^(2)` in the figure above. While they differ in their slope, i.e. in how much they charge per unit of effort (price per 1kb tx size), they are both reasonable pricing models.
   * I think for deciding which curve to pick, we should look at the throughput _limitations_ for some sample transactions. Roughly speaking, if we have two transactions A and B and we get twice as many transactions of type A through the system compared to transaction B, transaction B should cost twice as much. 
       * Formally, we require at least two benchmark transactions, A and B with their respective size `E_A` and `E_B`
       * We could load our benchmarking system with only these transactions up to the limit and measure the maximally achievable throughput. By comparing the throughput limitations, we can infer the slope. If we use more than 2 transactions, we get a sample of possible slopes. 

<img width="1218" alt="Screen Shot 2021-11-25 at 9 13 16 PM" src="https://user-images.githubusercontent.com/26322303/143530552-b09a9aa7-f29b-4beb-b00d-ac0c5189e382.png">